### PR TITLE
Refactor calendar engine into its own package

### PR DESCRIPTION
**Pull request by mickmister**
#### Summary

This pull requests refactors the main calendar engine used by the mscalendar plugin and leverages the remotes implementation to allow other plugins to easily import this plugin providing its own remote to easily create other calendar plugins.

This work was done to create the [mattermost-plugin-google-calendar](https://github.com/mattermost/mattermost-plugin-google-calendar) that could be used as a template, with the caveheat that the google calendar plugin contains it's own frontend code to create events and that is currently not shared and it's a task to tackle in the future.

### CODE REVIEW

**Comment by mickmister** [comment]
@fmartingr I've reopened this on a Mattermost-owned branch. Also notice that the base branch is `gcal-master`, which is currently equal to the `master` branch. 1/5 Let's use `gcal-master` as the protected branch for this project.

File: server/mscalendar/notification_test.go
Comment (mickmister) [inline server/mscalendar/notification_test.go:95]: Test needs to be reimplemented

File: server/command/info.go
Comment (mickmister) [inline server/command/info.go:15]: config: plugin name

File: server/config/const.go
Comment (mickmister) [inline server/config/const.go]: all fields being used here, that should be put into a struct

	BotUserName    = "gcal"
	BotDisplayName = "Google Calendar"
    BotDescription = "Created by the Google Calendar Plugin."
	ApplicationName    = "Google Calendar"
	Repository         = "mattermost-plugin-gcal"
	CommandTrigger     = "gcal"
	TelemetryShortName = "gcal"

File: server/mscalendar/settings_notifications.go
Comment (mickmister) [inline server/mscalendar/settings_notifications.go:40]: Not really sure what's going on here. It seems like:

"If a setting is a boolean, create an event subscription if not exists"

This is `notificationSetting.Set`, so I suppose this is "I want to change notification settings". what do we do when the value hasn't changed? I suppose this function takes care of no-ops

File: server/mscalendar/subscription.go
Comment (mickmister) [inline server/mscalendar/subscription.go:24]: Need to figure out if google's subscription lifecycle lines up with this

File: server/mscalendar/welcomer.go
Comment (mickmister) [inline server/mscalendar/welcomer.go:36]: "Microsoft Calendar plugin" mentioned here

File: server/mscalendar/welcomer.go
Comment (mickmister) [inline server/mscalendar/welcomer.go:128]: "Microsoft" mentioned here

File: server/plugin/plugin.go
Comment (mickmister) [inline server/plugin/plugin.go:179]: mscalendar terminology leaked into `plugin` package. not much else than what's visible here though

File: server/utils/oauth2connect/oauth2_complete.go
Comment (mickmister) [inline server/utils/oauth2connect/oauth2_complete.go:31]: "Microsoft Calendar" is hardcoded

**Review by mickmister — COMMENTED** [review]
@fmartingr I gave the plugin a review, and noted some things within this commit https://github.com/mattermost/mattermost-plugin-mscalendar/pull/267/commits/d78381aa8bf00887bab751a9d2a33d7571c6c5e5. Note that the commit itself is the review. Please take a look and share any thoughts you might have.

I added some PR review comments on top of the commit, to gather info that is provider-specific. The goal is to put them all in one struct. I plan to search for these strings in the code to find more:
- microsoft
- msgraph
- mscalender
- mscal
- gcal
- google

I think the notification subscription lifecycle is the one thing that seems murky with regards to remaining work to handle both providers. I need to better understand gcal's subscription lifecycle to connect the dots on the code related to subscriptions.

**Comment by mickmister** [comment]
Follow ups:

- make tickets for all known remaining tasks
- review original migration PR to make note of any significant changes for gcal case
- remove features/endpoints:
    - accept/decline/tentative
    - create/delete calendar

File: assets/profile.png
Comment (fmartingr) [inline assets/profile.png:1]: This we should take care as well.

**Comment by fmartingr** [comment]
> I need to better understand gcal's subscription lifecycle to connect the dots on the code related to subscriptions

Using the [Events.watch](https://developers.google.com/calendar/api/v3/reference/events/watch) endpoint to create a channel, with a default TTL of 7 days, channel can be stopped using [Channels.stop](https://developers.google.com/calendar/api/v3/reference/channels/stop).

**Comment by raghavaggarwal2308** [comment]
@mickmister While uploading the plugin we are getting the following error in the system console.
image (2)

I think the issue might be on [this](https://github.com/mattermost/mattermost-plugin-mscalendar/blob/8b4065b20bacfecf539a8091107ee2e14415a719/calendar/command/command.go#L84) line of the code. We are getting the provider name as "msgraph" but our assets folder has the files with the name "mscalendar". Do you want us to look into the issue or will you be looking into it? Also, please let us know if we might be missing out on something in the setup.

**Comment by fmartingr** [comment]
> @mickmister While uploading the plugin we are getting the following error in the system console. image (2)
> 
> I think the issue might be on [this](https://github.com/mattermost/mattermost-plugin-mscalendar/blob/8b4065b20bacfecf539a8091107ee2e14415a719/calendar/command/command.go#L84) line of the code. We are getting the provider name as "msgraph" but our assets folder has the files with the name "mscalendar". Do you want us to look into the issue or will you be looking into it? Also, please let us know if we might be missing out on something in the setup.

Let me do an initial setup in a local Mattermost instance to check that the basics are ok.

**Comment by fmartingr** [comment]
I've pushed a commit to use `mscalendar` as bot name/command. Profile pictures should work as well.

**Comment by mickmister** [comment]
Hi @AayushChaudhary0001, just checking in. How is it going with the testing of this plugin?

**Comment by AayushChaudhary0001** [comment]
**Tested this PR.**

2 commands, namely ``createevent`` and ``findmeetings`` do not have any functionality mentioned anywhere and is not clear after executing those commands as well.
Also, on adding any calendar through ``createcal``, I am not able to find from where to add it in the Outlook window. Can you please have a look at those two and this issue?

Rest all are tested and approved from my side.

**Comment by mickmister** [comment]
Thanks so much @AayushChaudhary0001. All 3 of those commands are generally unimplemented and should be removed from the codebase

**Comment by fmartingr** [comment]
> Thanks so much @AayushChaudhary0001. All 3 of those commands are generally unimplemented and should be removed from the codebase

Should I do that in a separate PR?

**Comment by mickmister** [comment]
@fmartingr Yeah that sounds good to me

**Comment by fmartingr** [comment]
> > Rest all are tested and approved from my side.
> 
> @fmartingr It seems this PR has been approved by QA. Should we merge this and move forward with a release of this plugin, and also update the gcal plugin's commit to the next tag once created?

Absolutely! You take care of mscalendar and I take care of gcal?

**Comment by fmartingr** [comment]
Solved conflicts from latest changes.

**Comment by mickmister** [comment]
@fmartingr Do you think this is good to merge based on QA's approval?

**Comment by fmartingr** [comment]
> @fmartingr Do you think this is good to merge based on QA's approval?

Yeah, apart from QA we did some testing ourselves back in the day, so I think this is good to go.

**Comment by mickmister** [comment]
@fmartingr I need at least one dev's explicit approval on this to merge it. Can you please take a look and approve if it looks good? (which it does based on your comment)

**Review by fmartingr — APPROVED** [review]
Considering this passed through QA and the Google Calendar plugin works from this, I'm giving my approval, but if we need a proper 2nd dev review I would add someone else, since we are the authors of this PR so my approval seems like cheating.

**Comment by mickmister** [comment]
@larkox Do you mind giving this a review?

**Review by larkox — APPROVED** [review]
Apart of the many "REVIEW" comments, everything seems fine. And I find this a great idea!

Are we planning on dealing with the review comments now or later?

**Comment by mickmister** [comment]
@larkox I created two followup issues from those "REVIEW" comments

https://github.com/mattermost/mattermost-plugin-mscalendar/issues/357
https://github.com/mattermost/mattermost-plugin-mscalendar/issues/358

**Pull request merged**
